central/auth/datastore/datastore_impl.go

```diff
@@ -3,43 +3,103 @@ package datastore
 import (
 	"context"
 
+	"github.com/stackrox/rox/central/auth/m2m"
 	"github.com/stackrox/rox/central/auth/store"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
 var (
 	_ DataStore = (*datastoreImpl)(nil)
+
+	accessSAC = sac.ForResource(resources.Access)
 )
 
 type datastoreImpl struct {
 	store store.Store
+	set   m2m.TokenExchangerSet
+
+	mutex sync.RWMutex
 }
 
 func (d *datastoreImpl) GetAuthM2MConfig(ctx context.Context, id string) (*storage.AuthMachineToMachineConfig, bool, error) {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	return d.getAuthM2MConfigNoLock(ctx, id)
+}
+
+func (d *datastoreImpl) getAuthM2MConfigNoLock(ctx context.Context, id string) (*storage.AuthMachineToMachineConfig, bool, error) {
 	return d.store.Get(ctx, id)
 }
 
 func (d *datastoreImpl) ListAuthM2MConfigs(ctx context.Context) ([]*storage.AuthMachineToMachineConfig, error) {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
 	return d.store.GetAll(ctx)
 }
 
 func (d *datastoreImpl) AddAuthM2MConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) (*storage.AuthMachineToMachineConfig, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
 	setIssuer(config)
+	exchanger, err := d.set.NewTokenExchangerFromConfig(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
 	if err := d.store.Upsert(ctx, config); err != nil {
 		return nil, err
 	}
+
+	if err := d.set.UpsertTokenExchanger(exchanger, config.GetIssuer()); err != nil {
+		return nil, err
+	}
 	return config, nil
 }
 
 func (d *datastoreImpl) UpdateAuthM2MConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) error {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
 	setIssuer(config)
+	exchanger, err := d.set.NewTokenExchangerFromConfig(ctx, config)
+	if err != nil {
+		return err
+	}
+
 	if err := d.store.Upsert(ctx, config); err != nil {
 		return err
 	}
+
+	if err := d.set.UpsertTokenExchanger(exchanger, config.GetIssuer()); err != nil {
+		return err
+	}
 	return nil
 }
```

Comment:
It seems to me more difficult than that. Consider the case there's already a config with the given `id` with some issuer `issuerA` and it gets updated to `issuerB`. We need to remove the old issuer from the set and add a new one.

I need to think more about the problem but wanted to quickly share it so that you can check this too.

Reply:
Yeah.
What I've done:
- Fetch the config before updates are applied.
- Do the whole creation of the new token exchanger.
- If the updated config's issuer has changed (i.e. existing.issuer != updated.issuer) remove the previous issuer from the set.

This should be it.